### PR TITLE
refactor combine logic for clarity

### LINKED ISSUE #1448: Combine requires a string array but the code handles it as an array of object (DEVX-83)
It is not possible to combine multiple signatures using the `api.combine` function.

```js
function combine(signedTransactions: Array<string>): object {
  validate.combine({signedTransactions})

  // TODO: signedTransactions is an array of strings in the documentation, but
  // tests and this code handle it as an array of objects. Fix!
  const txs: any[] = _.map(signedTransactions, binary.decode)
  const tx = _.omit(txs[0], 'Signers')
  if (!_.every(txs, (_tx) => _.isEqual(tx, _.omit(_tx, 'Signers')))) {
    throw new utils.common.errors.ValidationError(
      'txJSON is not the same for all signedTransactions'
    )
  }
  const unsortedSigners = _.reduce(
    txs,
    (accumulator, _tx) => accumulator.concat(_tx.Signers || []),
    []
  )
  const signers = unsortedSigners.sort(compareSigners)
  const signedTx = _.assign({}, tx, {Signers: signers})
  const signedTransaction = binary.encode(signedTx)
  const id = computeBinaryTransactionHash(signedTransaction)
  return {signedTransaction, id}
}
```

### PR DESCRIPTION
**Pull request by JST5000**
## High Level Overview of Change

Refactor api.combine to be more readable.

### Context of Change
Resolves #1448 

I refactored the combine method to clearly highlight the core steps that were required instead of making the reader infer the high level goals from a bunch of lambda functions.

### Type of Change

- [X] Refactor (non-breaking change that only restructures code)

## Before / After

Before it was hard to understand the point of each lambda function, now it should be easy to quickly find the part of the function that's causing problems or needs improving.

## Test Plan

CI Passes

### CODE REVIEW

File: src/transaction/combine.ts
Comment (mvadari) [inline src/transaction/combine.ts:21]: Import this directly from the module instead of using `import *` syntax

File: src/transaction/combine.ts
Comment (mvadari) [inline src/transaction/combine.ts:51]: This is bad practice - this will modify the actual object at `transactions[0]`. That's why the syntax in deleted line 38 exists - so that a new object is created for that (I know it's a little weird). It may not matter for how you're using it, but someone else might use the same method and not realize.

File: src/transaction/combine.ts
Comment (mvadari) [inline src/transaction/combine.ts:62]: Nice docs!

File: src/transaction/combine.ts
Comment (mvadari) [inline src/transaction/combine.ts:10]: Can you add these types of quick explaining comments to the other methods in this file as well?

**Review by mvadari — CHANGES_REQUESTED** [review]
Nice job!

File: src/transaction/combine.ts
Comment (natenichols) [inline src/transaction/combine.ts:4]: Can we adjust the spacing on these imports to be `{ValidationError}` to match the rest of the code. This should get fixed when we address linting, though.

File: src/transaction/combine.ts
Comment (natenichols) [inline src/transaction/combine.ts:8]: I _think_ this is the wrong JsonObject.. This is the one in ripple-binary-codec. Think theres a different one in ripple-lib, `TransactionJson`

Comment (natenichols) [reply src/transaction/combine.ts:8]: Really, we should put TransactionJson in the binary codec, and then import it into ripple-lib.

File: src/transaction/combine.ts
Comment (natenichols) [inline src/transaction/combine.ts:58]: Can we use `Object.assign()` here

File: src/transaction/combine.ts
Comment (natenichols) [inline src/transaction/combine.ts:73]: Ahhh, I see why you're using that here. Disregard my early comments.

File: src/transaction/combine.ts
Comment (natenichols) [inline src/transaction/combine.ts:49]: Can we use `transactions.reduce` here. I want to try to add as little lodash as possible, as we're planning to remove it.

File: src/transaction/combine.ts
Comment (natenichols) [reply src/transaction/combine.ts:73]: This is totally fine, then, especially as you're not exposing this type to the client.

Comment (natenichols) [reply src/transaction/combine.ts:73]: Frankly, I made bad design decisions w/ those types I would love to go back and improve how ripple binary codec does types

**Review by natenichols — APPROVED** [review]
Awesome work @JST5000. Left a couple requests for change, but nothing that is important enough to hold up approval.

LGTM 👍 (pending those changes)

File: src/transaction/combine.ts
Comment (mvadari) [reply src/transaction/combine.ts:73]: we're about to do that ;)

**Review by mvadari — APPROVED** [review]
lgtm! just make sure you do what Nathan's comments said, and you're good to go

**Review by intelliot — CHANGES_REQUESTED** [review]
This looks good. For the single-line comments with `//`, it would be ideal to put a space between the `//` and the text, as that's the way all the other comments in the project are styled (I believe). In the future, this is hopefully auto-fixable by the linter.

And as Nate mentioned, we would prefer to move away from using lodash `_` if possible.

File: src/transaction/combine.ts
Comment (ledhed2222) [inline src/transaction/combine.ts:16]: why the underscore before `_tx`? as a convention, a leading underscore should mean that you aren't using that argument in the function body, but in this case you are. please remove

File: src/transaction/combine.ts
Comment (ledhed2222) [inline src/transaction/combine.ts:18]: there's no reason to delete this since you are re-assigning it later. using `delete` is weird

File: src/transaction/combine.ts
Comment (ledhed2222) [inline src/transaction/combine.ts:20]: errant space inside paren

File: src/transaction/combine.ts
Comment (ledhed2222) [inline src/transaction/combine.ts:14]: no reason to delete this since you reassign it later

File: src/transaction/combine.ts
Comment (ledhed2222) [inline src/transaction/combine.ts:13]: leading underscore is confusing and indicates that the variable is not going to be used, but it will

File: src/transaction/combine.ts
Comment (ledhed2222) [inline src/transaction/combine.ts:11]: why these extra spaces inside parens?

File: src/transaction/combine.ts
Comment (ledhed2222) [reply src/transaction/combine.ts:10]: should have have a space after `//`

File: src/transaction/combine.ts
Comment (ledhed2222) [inline src/transaction/combine.ts:28]: overall it seems like this can be re-written much simpler like so:

```javascript
function validateAllSignedTransactionsAreEqual(transactions : Array<JsonObject>) {
  if (transactions.every(tx => JSON.stringify(tx) === JSON.stringify(transactions[0])) {
    return transactions
  }
  throw new ValidationError('txJSON is not the same for all signedTransactions')
}
```

File: src/transaction/combine.ts
Comment (ledhed2222) [reply src/transaction/combine.ts:18]: ah nevermind i see why you're doing this. but there's a better way - see my suggestion below!

File: src/transaction/combine.ts
Comment (ledhed2222) [inline src/transaction/combine.ts:54]: space after `//`

File: src/transaction/combine.ts
Comment (ledhed2222) [inline src/transaction/combine.ts:57]: space after `//`

File: src/transaction/combine.ts
Comment (ledhed2222) [reply src/transaction/combine.ts:10]: also, since you have a function that is validating this condition you can rewrite this comment to be more clear in terms of what the function is doing, as opposed to the abstract requirement. See my suggested function below.

File: src/transaction/combine.ts
Comment (ledhed2222) [reply src/transaction/combine.ts:49]: this is also better described as a `map`.

File: src/transaction/combine.ts
Comment (ledhed2222) [reply src/transaction/combine.ts:58]: there's no benefit to assigning `unsortedSigners` to a variable, since it will only be used once. I think that makes your code unnecessarily verbose. instead:

```javascript
function getATransactionWithAllSigners(transactions : Array<JsonObject>) : JsonObject {
  // Signers must be sorted
  const sortedSigners = transactions
    .map(tx => tx.Signers)
    .filter(signer => signer)
    .sort(compareSigners)

  return Object.assign({}, {...transactions[0], Signers: sortedSigners})
}
```

File: src/transaction/combine.ts
Comment (ledhed2222) [inline src/transaction/combine.ts:74]: if you're not using the return value of `transactions` then don't return anything from this. updated proposed function above.

**Review by ledhed2222 — CHANGES_REQUESTED** [review]
using functional styles will let you further simplify this :) see comments

File: src/transaction/combine.ts
Comment (JST5000) [reply src/transaction/combine.ts:28]: That's much better! 

The "...object" syntax is super useful for this use case :)

File: src/transaction/combine.ts
Comment (JST5000) [reply src/transaction/combine.ts:58]: That looks cleaner to me :) 
(Unfortunately I still had to use lodash to get flatMap as it isn't built in, but it's much more readable than assign!)

File: src/transaction/combine.ts
Comment (ledhed2222) [inline src/transaction/combine.ts:41]: check whether you can do `map` followed by `flat` if there's no polyfill for `flatMap`. otherwise, this is a good reason to use lodash or something similar

Comment (JST5000) [reply src/transaction/combine.ts:41]: "flat" is also not appearing in the polyfill unfortunately, so I think lodash is probably the best bet

File: src/transaction/combine.ts
Comment (ledhed2222) [inline src/transaction/combine.ts:62]: i don't think you gain readability from assigning these variables. i think this is actually clearer, especially because the keys of the return are descriptive

```javascript
validateAllSignedTransactionsAreEqual(transactions)

const signedTransaction = binary.encode(getATransactionWithAllSigners(transactions))
return {
  id: computeBinaryTransactionHash(signedTransaction),
  signedTransaction,
}
```

Comment (JST5000) [reply src/transaction/combine.ts:62]: Yep, I'd agree with that. The function name makes it clear what it's doing.

File: src/transaction/combine.ts
Comment (ledhed2222) [inline src/transaction/combine.ts:13]: you've got some very long function names that could be terser without losing detail. 

this could be `validateSignedTransactionEquality` (probably shorter still but a start)

File: src/transaction/combine.ts
Comment (ledhed2222) [inline src/transaction/combine.ts:37]: how about `firstTransactionWithAllSigners`

File: src/transaction/combine.ts
Comment (ledhed2222) [reply src/transaction/combine.ts:13]: actually - the `Signed` part is confusing, since we are validating transactions _regardless_ of signer. So it could just be `validateTransactionEquality`.

Comment (JST5000) [reply src/transaction/combine.ts]: Good point!

File: src/transaction/combine.ts
Comment (JST5000) [reply src/transaction/combine.ts:37]: I think 'getA' conveys the intent better than 'first' for this one.

File: src/transaction/combine.ts
Comment (JST5000) [reply src/transaction/combine.ts:41]: Checked out the compiler errors caused by the flatMap configuration change, but it seemed like it would take me a long time to figure out the root cause and debug it all. Didn't seem worth it for removing one instance of lodash in a file which already is using lodash elsewhere. (If you want to see the config change, you can see it in the commit c7f2d28)

Comment (ledhed2222) [reply src/transaction/combine.ts:41]: I mean I'm team lodash lol so I'm ok with this 😹

File: src/transaction/combine.ts
Comment (ledhed2222) [reply src/transaction/combine.ts:37]: there's something about the `ATransaction` in camel case that makes this hard to read. Is it "a transaction"? Or an "a-class transaction"? it looks strange to me. what about just `getTransactionWithAllSigners`

File: src/transaction/combine.ts
Comment (ledhed2222) [inline src/transaction/combine.ts:11]: i don't think "except" needs to be all-caps in this case. the all-caps implies that a caller could do something wrong if they don't do whatever the all-caps indicates, but in this case this is just what the function does, NBD to user.

File: src/transaction/combine.ts
Comment (ledhed2222) [inline src/transaction/combine.ts:29]: Link to the documentation and remove "According to the documentation"

File: src/transaction/combine.ts
Comment (ledhed2222) [reply src/transaction/combine.ts:11]: if you're in here already, "equivalence" would be even more descriptive than "equality".

File: src/transaction/combine.ts
Comment (mvadari) [reply src/transaction/combine.ts]: It's totally fine to use lodash's `flatMap`, it's used elsewhere too I think

File: src/transaction/combine.ts
Comment (mvadari) [reply src/transaction/combine.ts:37]: Yeah I like that, or `getTransactionAllSigners`

File: src/transaction/combine.ts
Comment (mvadari) [reply src/transaction/combine.ts:29]: +1

File: src/transaction/combine.ts
Comment (mvadari) [reply src/transaction/combine.ts:37]: You generally don't put articles (a, an, the) in method names, unless they're serving some sort of specific purpose

File: src/transaction/combine.ts
Comment (JST5000) [reply src/transaction/combine.ts:28]: Yep, that makes sense :)

**Review by intelliot — APPROVED** [review]
LGTM. Thanks for taking this on!

**Pull request merged**
